components/brave_extension/extension/brave_extension/content.ts

```diff
@@ -1,10 +1,100 @@
+// Notify the background script as soon as the content script has loaded.
+// chrome.tabs.insertCSS may sometimes fail to inject CSS in a newly navigated
+// page when using the chrome.webNavigation API.
+// See: https://bugs.chromium.org/p/chromium/issues/detail?id=331654#c15
+// The RenderView should always be ready when the content script begins, so
+// this message is used to trigger CSS insertion instead.
+chrome.runtime.sendMessage({
+  type: 'pageContentReadyForInjection'
+})
+
 const unique = require('unique-selector').default
+
 let target: EventTarget | null
+let genericExceptions: Array<string> | undefined = undefined
+let classIdBuffer: { classes: string[], ids: string[] } = { classes: [], ids: [] }
+
+const queriedIds = new Set()
+const queriedClasses = new Set()
+const regexWhitespace = /\s/
 
 function getCurrentURL () {
   return window.location.hostname
 }
 
+const getClassesAndIds = function (addedNodes: Element[]) {
```

Comment:
This is what uBlock Origin does and I haven't found the performance hit to be noticeable in practice, but it would be nice to take advantage of native APIs if possible.